src/Authenticator/LoginHandler.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace SilverStripe\RealMe\Authenticator;
+
+use Exception;
+use Psr\Log\LoggerInterface;
+use SilverStripe\Control\Controller;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\RequestHandler;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\RealMe\Exception as RealMeException;
+use SilverStripe\RealMe\Extension\MemberExtension;
+use SilverStripe\RealMe\RealMeService;
+use SilverStripe\Security\Member;
+use SilverStripe\Security\RequestAuthenticationHandler;
+use SilverStripe\Security\Security;
+
+class LoginHandler extends RequestHandler
+{
+    private static $dependencies = [
+        'service' => '%$' . RealMeService::class,
+    ];
+
+    /**
+     * @var array
+     */
+    private static $url_handlers = [
+        '' => 'login',
+    ];
+
+    /**
+     * @var array
+     * @config
+     */
+    private static $allowed_actions = [
+        'login',
+        'acs',
+    ];
+
+    /**
+     * @var string Called link on this handler
+     */
+    private $link;
+
+    /**
+     * @var RealMeService
+     */
+    protected $service;
+
+    /**
+     * @param string $link The URL to recreate this request handler
+     */
+    public function __construct($link)
+    {
+        $this->link = $link;
+        parent::__construct();
+    }
+
+    /**
+     * Return a link to this request handler.
+     * The link returned is supplied in the constructor
+     * @param null|string $action
+     * @return string
+     */
+    public function link($action = null)
+    {
+        if ($action) {
+            return Controller::join_links($this->link, $action);
+        }
+
+        return $this->link;
+    }
+
+    /**
+     * URL handler for the log-in screen
+     *
+     * @return array
+     */
+    public function login()
+    {
+        return [
+            'Form' => $this->loginForm(),
+        ];
+    }
+
+    public function loginForm()
+    {
+        return LoginForm::create($this, 'acs');
+    }
+
+    public function acs(HTTPRequest $request)
+    {
+        try {
+            $authenticated = $this->service->enforceLogin($request);
+
+            $session = $request->getSession();
+
+            if ($authenticated === true) {
+                $authData = $this->service->getAuthData();
+
+                // If more session vars are set here, they must be cleared in realmeLogout()
+                $session->set('RealMe.SessionData', serialize($authData));
```

Comment:
Reckon we could use JSON for encoding instead? This would prevent security audits from flagging possible serialisation security issues from us unserialising variable data

Reply:
I played around with this idea. It's possible, but I have to do some finagling to get everything to work correctly. Perhaps in the interests of not letting this PR get too out of hand I'll raise a separate issue for this?

Reply:
Perfect

Reply:
#25